### DISCUSSION TITLE
safety_margin_buffer for collision costs

### DISCUSSION THREAD

**Discussion by rjoomen** [tesseract-robotics/trajopt#318] [Q&A]
I understand from #160 that for collision costs it is recommended to set the safety_margin_buffer to 0. Could you please explain why, as I have not been able to find information on this. (For constraints it's explained multiple times, e.g. #258.)

**Comment by Levi-Armstrong** [comment]
The safety_margin_buffer should never be zero. This is because you still want to have an entry in the SQP problem when you are near a collision so it knows not to move in that direction. Before this was added it would in one solver step get out of collision then on the next solver step things like velocity smoothing would push it back into collision. Then it would just ping-pong back and forth. The buffer ensures that for some distance outside the contact threshold the sqp problem still has a gradient entry preventing this from happening.

**Reply by Levi-Armstrong** [reply]
I see that I made a distinction between cost and constraints, but now I would say that it should be used for both. In the case of collision as a cost it would serve the same purpose so it has a gradient for some distance past the contact threshold.

**Reply by Levi-Armstrong** [reply]
In an ideal world you would just calculate all collision object pairs distance and gradient, but this would be computational slow using the current contact checkers. I implemented a GPU collision checker a while back where this would be possible.

**Reply by rjoomen** [reply]
OK, so the [default here](https://github.com/tesseract-robotics/tesseract_planning/blob/master/tesseract_motion_planners/trajopt/include/tesseract_motion_planners/trajopt/trajopt_collision_config.h#L62-L64) should be changed to something like 0.05 to match the newest insights?

**Reply by Levi-Armstrong** [reply]
I would provably set it to 0.01 as the default
